libs/core/langchain_core/prompts/prompt.py

```diff
@@ -104,12 +105,24 @@ def pre_init_validation(cls, values: dict) -> Any:
             )
 
         if values["template_format"]:
+            # Collect nested partial variables from
+            # BasePromptTemplate instances in partial_variables
+            nested_partial_vars = {
+                key
+                for partial_var in values["partial_variables"].values()
+                if isinstance(partial_var, BasePromptTemplate)
+                for key in partial_var.partial_variables
+            }
+
+            # Filter template variables based on
+            # partial_variables and nested_partial_vars
             values["input_variables"] = [
```

Comment:
This is the only point I found a bit counterintuitive. Imo if the user passes `input_variables` to the `PromptTemplate` we should not overwrite them with values from the template. This complicates the logic as I had to make sure that the nested partial variables are correctly excluded, to not break backward compatibility.